blockio-api/src/System/FS/BlockIO/API.hs

```diff
@@ -170,3 +210,40 @@ hAdviseAll hbio h advice = hAdvise hbio h 0 0 advice -- len=0 implies until the
 -- present.
 hDropCacheAll :: HasBlockIO m h -> Handle h -> m ()
 hDropCacheAll hbio h = hAdviseAll hbio h AdviceDontNeed
+
+{-------------------------------------------------------------------------------
+  File locks
+-------------------------------------------------------------------------------}
+
+-- | A handle to a file locked using 'tryLockFile'.
+newtype LockFileHandle m = LockFileHandle {
+    -- | Release a file lock acquired using 'tryLockFile'.
+    hUnlock :: m ()
+  }
+
+tryLockFileIO :: HasFS IO HandleIO -> FsPath -> GHC.LockMode -> IO (Maybe (LockFileHandle IO))
+tryLockFileIO hfs fsp mode = do
+    fp <- FS.unsafeToFilePath hfs fsp -- shouldn't fail
+    rethrowFsErrorIO hfs fsp $
+      bracketOnError (GHC.openFile fp GHC.WriteMode) GHC.hClose $ \h -> do
+        bracketOnError (GHC.hTryLock h mode) (\_ -> GHC.hUnlock h) $ \b -> do
+          if b then
+            pure $ Just LockFileHandle { hUnlock = rethrowFsErrorIO hfs fsp $ do
+                  GHC.hUnlock h
+                    `finally` GHC.hClose h
+                }
+          else
+            pure $ Nothing
+
+-- TODO: this is copied/adapted from System.FS.IO. Is there an alternative to
+-- copying this code?
+rethrowFsErrorIO :: HasCallStack => HasFS IO HandleIO -> FsPath -> IO a -> IO a
+rethrowFsErrorIO hfs fp action = do
+    res <- try action
+    case res of
+      Left err -> handleError err
+      Right a  -> return a
+  where
+    handleError :: HasCallStack => IOError -> IO a
+    handleError ioErr =
+      throwIO $ FS.ioToFsError (FS.mkFsErrorPath hfs fp) ioErr
```

Comment:
It's not exported there, so not sure how to avoid duplicating it. Also, I wonder why it's not just doing ``action `catch` handleError``.

Reply:
Good question, I don't know :stuck_out_tongue: